### ISSUE TITLE
OnClick Effect is not working on keyboard events

### ISSUE THREAD

**Issue by joliver82** [nifty-gui/nifty-gui#452]
I'm working with jme3 and niftygui and I've found out that if pressing enter or space keys when having focus on a nifty element with an interact tag the onclick event is not run although according to the manual it should: “OnClick: The element has been clicked by the mouse or is activated by a keyboard interaction”.

The interaction method defined is properly called in any case (clicking or pressing keys).

I've been working with a simple test case (posted at jme3 fourm https://hub.jmonkeyengine.org/t/niftygui-onclick-effect-is-not-working-on-keyboard-events/40572) and realized that the onclick effect is run from PrimaryClickMouseMethods.onClick method and not from any other place.

Keyboard events follow a different path until ElementInteractionClickHandler or MouseClickMethods clickAndRelease methods so I've though that modifying de.lessvoid.nifty.elements.ElementInteractionClickHandler or de.lessvoid.nifty.elements.MouseClickMethods to make the effect run would suffice.

Either one of the following:

de.lessvoid.nifty.elements.ElementInteractionClickHandler

```
  public void clickAndReleaseMouse(@Nonnull final Nifty nifty) {
    element.startEffectWithoutChildren(EffectEventId.onClick); //new line
    mouseMethods.clickAndRelease(nifty);
  }
```

de.lessvoid.nifty.elements.MouseClickMethods

```
  public void clickAndRelease(@Nonnull final Nifty nifty) {
    element.startEffectWithoutChildren(EffectEventId.onClick); //new line
    if (onClickMethod != null) {
      onClickMethod.invoke();
    }
    if (onReleaseMethod != null) {
      onReleaseMethod.invoke();
    }
  }
```

But I'm not sure if there could be an issue of running twice the effect in case both onClick and ClickAndRelease are called from mouse interaction (tried it and didn't happen but just in case asking...)

If it seems fine from your point of view I can create a PR if you prefer

Thanks

**Comment by bgroenks96** [comment]
Hi, sorry for the delayed response.

I reviewed the documentation and the relevant code. I agree that, in the current language of the manual, it should be the case that the `OnClick` *effect* is started by key events dispatched to an element. This is, however, rather confusing, as the normal `onClick` interaction handler *should not* be activated by key events. I think it would be ideal to have a separate effect for key events rather than handling them in `OnClick`, but this would require updating the documentation and could end up also bringing about breaking changes, so I would rather target compliance with the current documentation in this case.

Now, it's been a long time since I've actually stepped through any of this code, so I could be mistaken, but I don't think your proposed changes will do what you want. If you've tested it already and it works, please feel free to correct me (and submit a PR so that I can reproduce it). But a quick glance over the relevant parts of the `Element` class (which is responsible for dispatching these events) tells me that only real mouse click events are directed to `ElementInteractionClickHandler` and `MouseClickMethods`. This is, of course, by design. Keyboard events are directed to the element's key event handlers. It might make more sense to simply trigger the effect in the element class where the keyboard events are initially handled.

Maybe @void256 could weigh in on this as well. Mucking around with the input handling system can be... tricky, to say the least :wink:

**Comment by joliver82** [comment]
Hi

Sorry for the late reply too, I was on vacacion the whole month 

I agree that it's confusing having both keyboard and mouse events handled in the same onClick tags (either interacts or effects) but this is how it's defined and the way it has been working forever at least from when I started using nifty

About the difference between keyboard and mouse events, the "problem" is that default controllers like MenuItemControl translate some generic events into mouse events, in this case the NiftyStandardInputEvent.Activate event being enter or space keys:

```
public class MenuItemControl extends AbstractController {
...............
  @Override
  public boolean inputEvent(@Nonnull final NiftyInputEvent inputEvent) {
    Element element = getElement();
    if (element == null) {
      return false;
    }
...............
    } else if (inputEvent == NiftyStandardInputEvent.Activate) {
      element.onClickAndReleasePrimaryMouseButton();
      return true;
    }
    return false;
  }
```

So when enter or space key has been pressed (activate event), the item controlled changes this event into mouse click and release while real mouse clicks end up calling both onClick methods and onRelease methods but never onClickAndRelease (or at least I've not found the case)

I tested it before posting this issue and it worked properly in my environment (jme3+nifty) so I'm creating a PR for you to test it ;)

**Comment by bgroenks96** [comment]
Fixed by #454
